### PR TITLE
Allow users to reference datastore items inside the rule criteria pattern

### PR DESCRIPTION
**Pull request by Kami**
This pull request allows users to access datastore items inside a rule criteria pattern by utilizing a jinja variable expression.

Before, you could only reference a datastore item inside the criteria key, but not inside the pattern value - imo, value makes a lot more sense than criterion key.

@DoriftoShoes had a good use case for this one and I also have one.

For example:

```yaml
...

    criteria:
        trigger.parameters.cmd:
            type: "equals"
            pattern: "{{ system.action_cmd }}"
...
```

TODO:

- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by DoriftoShoes** [comment]
woohoo!  Thanks buddy.

File: st2reactor/st2reactor/rules/filter.py [hunk 1]
```diff
@@ -85,19 +109,33 @@ def _check_criterion(self, criterion_k, criterion_v, payload_lookup):
...
             return False
 
+        return result
+
+    def _render_criteria_pattern(self, criteria_pattern):
+        if not criteria_pattern:
+            return None
+
+        if not isinstance(criteria_pattern, six.string_types):
+            # We only perform rendering if value is a string - rendering a string value makes no
```
Comment (lakshmi-kannan) [inline st2reactor/st2reactor/rules/filter.py]: This comment also makes no sense :P

Comment (Kami) [inline st2reactor/st2reactor/rules/filter.py]: Yeah, there is something missing :P

File: st2common/st2common/util/templating.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    :param context: Template context.
+    :type context: ``dict``
+    """
+    assert isinstance(value, six.string_types)
+    context = context or {}
+
+    env = Environment(undefined=StrictUndefined)
+    template = env.from_string(value)
+    rendered = template.render(context)
```
Comment (lakshmi-kannan) [inline st2common/st2common/util/templating.py]: We might have to wrap this within try catch.

Comment (Kami) [inline st2common/st2common/util/templating.py]: We do this inside the calling function (inside RuleFilter).

Our goal should always be to propagate exceptions at the low level and handle them higher up where it's most appropriate.

Comment (lakshmi-kannan) [inline st2common/st2common/util/templating.py]: Got it.

File: st2reactor/st2reactor/rules/filter.py [hunk 3]
```diff
@@ -41,8 +43,22 @@ def __init__(self, trigger_instance, trigger, rule):
...
+
     def filter(self):
-        LOG.info('Validating rule %s for %s.', self.rule.id, self.trigger['name'])
+        """
+        Return true if the rule is applicable to the provided trigger instance.
+
+        :rtype: ``bool``
+        """
+        LOG.info('Validating rule %s for %s.', self.rule.id, self.trigger['name'],
+                 extra=self._base_logger_context)
```
Comment (manasdk) [inline st2reactor/st2reactor/rules/filter.py]: putting the entire rule, trigger and triggerinstance object will be excessive.

File: st2reactor/st2reactor/rules/filter.py [hunk 4]
```diff
@@ -54,7 +70,9 @@ def filter(self):
 
         payload_lookup = PayloadLookup(self.trigger_instance.payload)
 
-        LOG.debug('Trigger payload: %s', self.trigger_instance.payload)
+        LOG.debug('Trigger payload: %s', self.trigger_instance.payload,
```
Comment (manasdk) [inline st2reactor/st2reactor/rules/filter.py]: This is pointless if you have the previous line at info

File: st2reactor/st2reactor/rules/filter.py [hunk 5]
```diff
@@ -62,20 +80,26 @@ def filter(self):
...
-        if 'pattern' not in criterion_v:
-            criterion_v['pattern'] = None
+        criteria_operator = criterion_v['type']
+        criteria_pattern = criterion_v.get('pattern', None)
+
+        # Render the pattern (it can contain a jinja expressions)
+        try:
+            criteria_pattern = self._render_criteria_pattern(criteria_pattern=criteria_pattern)
+        except Exception:
+            LOG.exception('Failed to render pattern value for key %s' % (criterion_k),
```
Comment (manasdk) [inline st2reactor/st2reactor/rules/filter.py]: include the pattern as well. Easier for debugging.

File: st2reactor/st2reactor/rules/filter.py [hunk 6]
```diff
@@ -85,19 +109,33 @@ def _check_criterion(self, criterion_k, criterion_v, payload_lookup):
...
+                          extra=self._base_logger_context)
             return False
 
-        criteria_pattern = criterion_v['pattern']
-
         op_func = criteria_operators.get_operator(criteria_operator)
 
         try:
-            return op_func(value=payload_value, criteria_pattern=criteria_pattern)
+            result = op_func(value=payload_value, criteria_pattern=criteria_pattern)
```
Comment (manasdk) [inline st2reactor/st2reactor/rules/filter.py]: is there any difference with this and previous impl?

**Comment by manasdk** [comment]
+1

File: st2reactor/st2reactor/rules/filter.py (hunk 4 above)
Comment (Kami) [inline st2reactor/st2reactor/rules/filter.py]: True, I didn't really touch the log level though, I just added additional info to the context.

File: st2reactor/st2reactor/rules/filter.py (hunk 6 above)
Comment (Kami) [inline st2reactor/st2reactor/rules/filter.py]: Nope, just a style thing - same end result / outcome.

File: st2reactor/st2reactor/rules/filter.py (hunk 5 above)
Comment (Kami) [inline st2reactor/st2reactor/rules/filter.py]: Agreed. Done in ea3f6e4b507494730c7c1be65ac66b8576441703.

**Pull request merged**